server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,314 @@
+import logging
+import statistics
+from collections import defaultdict
+from typing import Dict, Iterable, List, NamedTuple, Tuple
+
+from sortedcontainers import SortedList
+
+from ...config import config
+from ...decorators import with_logger
+from ..search import CombinedSearch, Match, Search, get_average_rating
+from .matchmaker import Matchmaker
+
+
+class GameCandidate(NamedTuple):
+    """
+    Holds the participating searches and a quality rating for a potential game
+    from the matchmaker. The quality is not the trueskill quality!
+    """
+    match: Match
+    quality: float
+
+
+class Container:
+    def __init__(self, rating_difference, content):
+        self.rating = rating_difference
+        self.content = content
+
+
+class UnevenTeamsException(Exception):
+    pass
+
+
+class NotEnoughPlayersException(Exception):
+    pass
+
+
+@with_logger
+class TeamMatchMaker(Matchmaker):
+    """
+    Matchmaker for teams of varied size. Untested for higher than 4v4 but it should work
+
+    Overview of the algorithm:
+    1. list all the parties in queue by their average rating.
+    2. Take a party and select the neighboring parties alternating between lower
+    and higher until you have 8 players. If the next party to select would leave
+    you with more than 8 players you skip that party and try the next one.
+    3. you now have a list of parties that will be in your potential game. Distribute
+     them into two teams.
+     Start with the easiest cases: one party makes a full team already or one party
+     has n-1 players so you only need to find the best fitting single player.
+     If that is not the case perform the karmarkar-karp algorithm to get a good approximation for a partition
+    4. add this game to a games list
+    5. repeat 2. to 4. for every party.
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
+    """
+    def find(self, searches: Iterable[Search], team_size: int) -> Tuple[List[Match], List[Search]]:
+        if not searches:
+            return [], []
+
+        self._logger.debug("========= starting matching algorithm =========")
+
+        searches = SortedList(searches, key=lambda s: s.average_rating)
```

Comment:
If we wanted to fully utilize trueskill we would have treat the rating distributions as actual distributions when performing mathematical operations on them (like getting the average of the team). When picking a singular value it is always arbitrary to a degree.
Trueskill assumes that a new player is somewhat average. This is a good assumption in shooters, because they all play more or less the same, but in RTS starting players are *really* bad, because you need to keep up with your economy or you autolose and the ai games don't teach this at all.
We noticed this and coded this interpolation to counteract this problem. Subtracting the deviation is my way to make this interpolation. I think this has the benefit that it utilizes trueskill to determine how fast this interpolation should go away. So it automatically happens slower with games that provide little info about the players' skills.

It could be that 3 * dev is too much. Maybe 2 or 2.5 is better. We should probably look at some player data to see how their winrate for their first games is to get a feeling how big that factor should be.

I already explained this to someone else in the past, so maybe we should add this rationale somewhere in the comments?